### PR TITLE
Add CRUD testing framework for fields

### LINKED ISSUE #2242: Decimal fields resets
# Bug report

## Describe the bug

In a list, I have 2 decimal fields. In Admin UI, updating one of those decimal fields makes them reset (both or one of them only). Sometimes it deletes both data, sometimes just one of them.

## To Reproduce

Steps to reproduce the behaviour. Please provide code snippets or a repository:

1. Create a new project `yarn create keystone-app bug-example`, when the wizard asks, name it `bug-example`, `blank` template and `mongoose` for database connection
2. Create 2 lists like the ones below (Person and Bill)
3. Create a new Bill with a Name, Price and Installment Price in decimal (with dot, like 1234.00 and 123.00). Try to make it in one go.
4. After creating hit Save and you'll notice the fields being resetted.
5. If you try to put the values there again it will keep resetting.

The code. Paste below `new Keystone`...

```javascript

const PersonSchema = {
  fields: {
    name: { type: Text, isRequired: true },
    shortName: { type: Text },
    cpf: { type: Text },
    email: { type: Text, isUnique: true, isRequired: true },
    telephone: { type: Text },
    billList: { type: Relationship, ref: 'Bill.person' }
  }
}

const BillSchema = {
  fields: {
    name: { type: Text, isRequired: true },
    description: { type: Text },
    price: { type: Decimal, isRequired: true },
    dueDate: { type: DateTime },
    barCode: { type: Text },
    installments: { type: Integer },
    installmentsPaid: { type: Integer },
    installmentsPrice: { type: Decimal },
    isPaid: { type: Checkbox },
    person: { type: Relationship, ref: 'Person.billList', many: true }
  }
}

keystone.createList('Person', PersonSchema)
keystone.createList('Bill', BillSchema)
```

## Expected behaviour

The values should stay put. In GraphQL and directly manipulating in the database works correctly.

## Screenshots

WZ9iYfbuI3

## System information

- OS: Windows 10 
- Browser (if applies): Firefox or Chrome
- Mongoose for Windows

### PR DESCRIPTION
**Pull request by jesstelford**
In doing so, I discovered (and fixed) a bug with the `Decimal` field which would be set to `null` if you were to update any _other_ field.

Fixes #2242

PR is best viewed [with whitespace changes disabled](https://github.com/keystonejs/keystone/pull/2522/files?w=1)

### CODE REVIEW

File: api-tests/fields.test.js [hunk 1]
```diff
@@ -7,40 +7,52 @@ const { multiAdapterRunners, setupServer } = require('@keystonejs/test-utils');
...
+          );
 
-          mod.filterTests(keystoneTestWrapper);
+        describe(`${mod.name} field`, () => {
+          if (mod.crudTests) {
+            describe(`CRUD operations`, () => {
+              mod.crudTests(keystoneTestWrapper);
+            });
+          } else {
+            test.todo('CRUD operations - tests missing');
```
Comment (jesstelford) [inline api-tests/fields.test.js]: Gives nice output when running the tests (without failing the suite):

Screen Shot 2020-03-16 at 12 28 54 am

File: package.json [hunk 2]
```diff
@@ -197,7 +197,7 @@
       "!packages/create-keystone-app/templates/**",
       "!packages/core/tests/default-entry/index.js",
       "!packages/fields/**/views/**",
-      "!packages/fields/**/filterTests.js",
+      "!packages/fields/**/test-fixtures.js",
```
Comment (jesstelford) [inline package.json]: Why the file renames? In adding the `crudTests` I realised the filename was misleading, so picked one I thought was general enough while still conveying the overall meaning.

Happy to be challenged on this one!

File: packages/fields-mongoid/src/Implementation.js [hunk 3]
```diff
@@ -92,15 +92,22 @@ export class KnexMongoIdInterface extends KnexFieldAdapter {
...
+        if (typeof item[this.path] === 'string') {
+          item[this.path] = normaliseValue(item[this.path]);
+        } else {
+          // Should have been caught by the validator??
+          throw `Invalid MongoID value given for '${this.path}'`;
+        }
       } else {
-        // Should have been caught by the validator??
-        throw `Invalid value given for '${this.path}'`;
+        item[this.path] = null;
```
Comment (jesstelford) [inline packages/fields-mongoid/src/Implementation.js]: After discovering the bug in the `Decimal` field, I went back and rewrote the other fields which do the same logic to make it a bit clearer (to my brain). This should be functionally identical to what it was before.

**Comment by Vultraz** [comment]
The decimal bug sounds like #2242

**Comment by jesstelford** [comment]
Am breaking this up into smaller PRs

**Pull request closed without merging**